indexing-service/src/main/java/org/apache/druid/indexing/compact/MSQCompactionJobTemplate.java

```diff
@@ -0,0 +1,191 @@
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Compaction template that creates MSQ SQL jobs using a templatized SQL with
+ * variables of the format {@code ${variableName}} for fields such as datasource
+ * name and start timestamp.
+ * <p>
+ * Compaction is triggered for an interval only if the current compaction state
+ * of the underlying segments DOES NOT match with the {@link #targetState}.
+ */
+public class MSQCompactionJobTemplate implements CompactionJobTemplate
+{
+  public static final String TYPE = "compactMsq";
+
+  public static final String VAR_DATASOURCE = "${dataSource}";
+  public static final String VAR_START_TIMESTAMP = "${startTimestamp}";
+  public static final String VAR_END_TIMESTAMP = "${endTimestamp}";
+
+  private static final DateTimeFormatter TIMESTAMP_FORMATTER =
+      DateTimeFormat.forPattern("yyyy-MM-dd HH:mm:ss.SSS");
+
+  private final ClientSqlQuery sqlTemplate;
+  private final CompactionStateMatcher targetState;
+
+  @JsonCreator
+  public MSQCompactionJobTemplate(
+      @JsonProperty("sqlTemplate") ClientSqlQuery sqlTemplate,
+      @JsonProperty("targetState") CompactionStateMatcher targetState
+  )
+  {
+    this.sqlTemplate = sqlTemplate;
+    this.targetState = targetState;
+  }
+
+  @JsonProperty
+  public ClientSqlQuery getSqlTemplate()
+  {
+    return sqlTemplate;
+  }
+
+  @JsonProperty
+  public CompactionStateMatcher getTargetState()
+  {
+    return targetState;
+  }
+
+  @Nullable
+  @Override
+  public Granularity getSegmentGranularity()
+  {
+    return targetState.getSegmentGranularity();
+  }
+
+  @Override
+  public List<CompactionJob> createCompactionJobs(
+      DruidInputSource source,
+      CompactionJobParams jobParams
+  )
+  {
+    final String dataSource = source.getDataSource();
+
+    // Identify the compactible candidate segments
+    final CompactionConfigBasedJobTemplate delegate =
+        CompactionConfigBasedJobTemplate.create(dataSource, targetState);
+    final DataSourceCompactibleSegmentIterator candidateIterator =
+        delegate.getCompactibleCandidates(source, jobParams);
+
+    // Create MSQ jobs for each candidate by interpolating the template variables
+    final List<CompactionJob> jobs = new ArrayList<>();
+    while (candidateIterator.hasNext()) {
+      final CompactionCandidate candidate = candidateIterator.next();
+      jobs.add(
+          new CompactionJob(
+              createQueryForJob(dataSource, candidate.getCompactionInterval()),
+              candidate,
+              CompactionSlotManager.getMaxTaskSlotsForMSQCompactionTask(sqlTemplate.getContext())
+          )
+      );
+    }
+
+    return jobs;
+  }
+
+  private ClientSqlQuery createQueryForJob(String dataSource, Interval compactionInterval)
```

Comment:
wondering aloud on if there is a way to make the query and formatting more flexible/extensible if folks want to use more than these standard format vars. Perhaps if someone wanted to do some filtering during compaction? Or would you suggest that if a person desires that, that it would be in some custom template type that they roll on their own?

Reply:
Yes, the filter can simply be included in the SQL.

> But if someone changed the query after it existed for some time that filter wouldn't be re-applied to already compacted segments since it is not persisted in the target state.

Yes, that is by design. We don't want compaction to be triggered as long as the compaction state has not changed.
This ensures that every minor tweak made to the SQL query does not end up recompacting everything.